### ISSUE TITLE
Have in-game consoles be programmable

### ISSUE THREAD

**Issue by moonheart08** [griefly/griefly#486]
This is one of those things everyone wanted, but could never pull off in BYOND

My personal suggestion is a RISC-V based system that has the feature set IMC (Base Integer instructions, Multiply/Divide instructions, Compact instructions) and has access to either 8k or 16k of RAM, and supports 32K floppies. 

VRAM will need to be 3200 bytes large in order to fit a 20x80 terminal, but a 20x40 would be more fitting for this, so we can say it's 1600 bytes. This byte count is based on 2 byte characters, one byte for color, and one for the character itself. This is of course only if the terminal can be written anywhere, and in that case putting a cap of 160 writes to VRAM per frame (with the screen at 24 frames) will only use 3840 bytes/sec of transfer to the client when all 160 writes are used per frame. 

The CPU itself should run at a reasonable speed that allows large numbers of them to be simulated server side without too much of an issue, preferably on a separate thread or two to prevent too many consoles from bogging the server into oblivion. Either 6kHZ or 12kHz would be appropriate. 

This could also allow an easy replacement for the lesser known NTSCRIPT seen on some servers to manipulate the comms, just make it a console program instead.

PDAs could *possibly* also be programmable, with much simpler specs, but large numbers of them can exist at any one time, more than there are consoles, so it might not be recommendable.

This is of course, a long term goal, and shouldn't be done any time soon (Atmos isn't even finished!)

**Comment by kremius** [comment]
It is a such kind of feature which everyone wants, but noone actually will use (sadly, I would like to change it).

The comms thingy wasn't used much when I played and actually I struggle with imagining some funny thing to do with it. Remember, it is the game and games should be fun. Games aren't fun if they are unbalanced. In today world all coding & software stuff ridiculously imbalanced, so just copying everything from it would be unwise.

One option to make coding somewhat useful and not imbalanced is to allow code firstly in some less global things - like doors. So like in order to hack door it will be needed to somehow understand code or change a firmware of a door, etc. It will be 1) not too overpowered 2) pretty funny 3) not obligatory.
Global things are much harder to balance - but it would be interesting to hear some suggestions about a coding related gameplay.

Another thing about asm-like coding: it is funny in some way but the amount of people who will be able to do at least something will drop drastically (and this amount wasn't really high in the first place). It would be pain to do anything big. It can be suitable for small things like doors, tho.

TL;DR: really primitive microcontrollers, not CPUs.

**Comment by mechmind** [comment]
For another option there is a WebAssembly virtual machine. Initially built for web but with general usefulness in mind it has some desirable features:
- Secure and portable sandbox environment, suitable for running code, generated from memory-unsafe languages like C;
- Embeddability: there are simple interface for interacting with "outside" world, runtime allows limits on memory and execution time;
- There are toolchains to compile C, C++, Go, Rust to wasm; scripting languages can be used by compiling their interpreters into wasm (and this method is not horribly underperfomant)
- JIT is coming, so interpreters will be even faster;
- Text assembly format for those who want maximum hardcore expirience or for examining/debugging generated code;
- Nice tooling;
- Rapidly growing community due to web already adopting this tech. Also big companies like Mozilla and Microsoft and Google and Apple support wasm.
Considering those points WebAssembly is viable piece of technology worth at least examining.

**Comment by kremius** [comment]
@mechmind IMHO, it's more suitable for scripting #170
